### PR TITLE
BCDA-3781 - RFC for service date filter

### PR DESCRIPTION
**Pull request by mtrang1263**
### Change Details

Adding RFC that describes approach for the service date filter.

### Acceptance Validation

Review RFC and sign off on the approach

### External References

- [BCDA-3781](https://jira.cms.gov/browse/BCDA-3781)

### Security Implications

- [ ] new software dependencies
- [ ] altered security controls
- [ ] new data stored or transmitted
- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

File: rfcs/0007-service-date-filter.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    - [Proposed Solution: Drawbacks](#proposed-solution-drawbacks)
+    - [Proposed Solution: Notable Alternatives](#proposed-solution-notable-alternatives)
+  - [Prior Art](#prior-art)
+  - [Future Possibilities](#future-possibilities)
+  - [Addendums](#addendums)
+
+## Motivation
+[Motivation]: #motivation
+
+BCDA needs to ensure that they are not returning beneficiary data to an ACO to which that beneficiary is not attributed. There are a variety of use-cases where this could occur if not handled properly:
```
Comment (msnook) [inline rfcs/0007-service-date-filter.md:30]: Should we list ACO Termination here too?

File: rfcs/0007-service-date-filter.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
+# RFC Proposal
+[RFC Proposal]: #rfc-proposal
+
+* RFC Proposal ID: `0007-service-date-filter` (fill me in with a unique ident)
```
Comment (msnook) [inline rfcs/0007-service-date-filter.md:4]: Can we remove the `(fill me in with a unique ident)`?

File: rfcs/0007-service-date-filter.md (hunk 1 above)
Comment (mtrang1263) [reply rfcs/0007-service-date-filter.md:30]: I wast thinking that would be a "Future" use case since there may be some nuances with that.

**Review by msnook — APPROVED** [review]
This looks good to me!

File: rfcs/0007-service-date-filter.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+### Proposed Solution: Unresolved Questions
+[Proposed Solution: Unresolved Questions]: #proposed-solution-unresolved-questions
+
+1. The `serviceEndDate` filter was previously implemented, but removed due to lack of adoption. While BCDA has identified a use case for this functionality, it may be BCDA-specific and thus should be implemented on the BCDA side.
+
+### Proposed Solution: Drawbacks
+[Proposed Solution: Drawbacks]: #proposed-solution-drawbacks
+
+1. Since the filtering is on the application layer, instead of the database layer, we may roundtrip claim entities that will be filtered out.
```
Comment (jzulim) [inline rfcs/0007-service-date-filter.md:82]: I'm also curious to see if there's any significant performance change with the additional filter. BFD ops can help you load test a PR when we get to that point.

File: rfcs/0007-service-date-filter.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  * Returns claims data that occurred before 2020-01-01
+* https://\<hostname\>/v1/fhir/ExplanationOfBenefit?patient=\<beneficiaryId\>&serviceEndDate=ge2020-01-01&serviceEndDate=le2020-01-31
+  * Returns claims data that occurred between 2020-01-01 and 2020-01-31 (inclusive)
+
+If a claim entity's date field is unset/null and the caller supplies a `serviceEndDate`, then the claim **will not** be returned to the caller. Since we cannot guarantee that the claim falls within the supplied filter, we cannot return it.
+
+### Proposed Solution: Unresolved Questions
+[Proposed Solution: Unresolved Questions]: #proposed-solution-unresolved-questions
+
+1. The `serviceEndDate` filter was previously implemented, but removed due to lack of adoption. While BCDA has identified a use case for this functionality, it may be BCDA-specific and thus should be implemented on the BCDA side.
```
Comment (jzulim) [inline rfcs/0007-service-date-filter.md:77]: In general, BFD is interested in being the gatekeeper of data, so filtering on the BFD side makes sense for us.

File: rfcs/0007-service-date-filter.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+BCDA needs to ensure that they are not returning beneficiary data to an ACO to which that beneficiary is not attributed. There are a variety of use-cases where this could occur if not handled properly:
+* End of year reassignment/attribution changes
+* Runouts
+
+## Proposed Solution
+[Proposed Solution]: #proposed-solution
+
+The solution will allow a client to specify a new query parameter (`serviceEndDate`) when searching EOB resources. When specified, the search filters resources against the passed in date range. 
+
+The table below captures the claim type field used in the serviceEndDate filter.
```
Comment (jzulim) [inline rfcs/0007-service-date-filter.md:39]: 👍

**Review by jzulim — DISMISSED** [review]
Thank you for putting this together. Looks like a reasonable approach to me, the fields match with what I would expect to filter on for each claim type. BFD wants to be the gatekeepers of data released so for me, this belongs on BFD. BFD ops can help you do a performance test to see if the additional filtering has any significant performance hit.

File: rfcs/0007-service-date-filter.md (hunk 3 above)
Comment (mtrang1263) [reply rfcs/0007-service-date-filter.md:82]: Good to know John! I'll note in the implementation PR to make sure we have some numbers

File: rfcs/0007-service-date-filter.md (hunk 4 above)
Comment (dshekhar18) [reply rfcs/0007-service-date-filter.md:77]: @mtrang1263 

**serviceEndDate** is not a EOB standard search param in FHIR R4 and CARIN. You can use HTTP header to pass serviceEndDate.

OR

Use the below CARIN defined EOB search param called service-date:

GET [base]/ExplanationOfBenefit?service-date=[service-date] and you can pass ExplanationOfBenefit.billablePeriod.end or ExplanationOfBenefit.item.servicedDate

**Review by dshekhar18 — COMMENTED** [review]
https://github.com/CMSgov/beneficiary-fhir-data/pull/376#discussion_r510222956

File: rfcs/0007-service-date-filter.md (hunk 4 above)
Comment (mtrang1263) [reply rfcs/0007-service-date-filter.md:77]: Didn't know there was a standard search param for service date, thanks for sharing @dshekhar18 ! 

Will update to use `service-date` instead.

**Review by dshekhar18 — APPROVED** [review]
EOB search looks conformant with R4 and CARIN BB. Please deploy to BFD TEST as a POC exercise during implementation phase.

**Review by jzulim — APPROVED** [review]
RFC seems flushed out and in line with FHIR standards (i.e. service-date). The fields you've identified as fields to incorporate into the search parameters per claim type are all listed as mandatory so in theory this should pull back all claims needed. Approved.

**Pull request merged**